### PR TITLE
First attempt at parallel diagnostics

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
I should also thank @DrTodd13 for providing extremely valuable feedback whilst writing this.

**Comment by ehsantn** [comment]
This is great!

Printing what operation generated the loop would also be useful. Currently, `arrayexprs` are counted as a single loop. For example, the pi example has this line:
```
x = 2*np.random.ranf(n)-1---------------------------------------------------------| #0, 1
```
Also, some lines could have several parfor and non-parfor operations. For example, logistic regression code has this line (the outer `dot` is not parallelized but the inner one is):
```
w -= np.dot(((1.0 / (1.0 + np.exp(-Y * np.dot(X, w))) - 1.0) * Y), X)------------| #0, 1, 2
```

**Comment by stuartarchibald** [comment]
Thanks taking a look and the feedback @ehsantn.

If you turn up the verbosity on the diagnostic output (max is 4), does that show what you want? For example, with `level=4` on the `pi` example, I see this:
```
================================================================================
======= Parallel Accelerator Optimizing:  Function calc_pi, pi.py (13)  ========
================================================================================
---------------------------Looking for parallel loops---------------------------
Parallel for-loop #0: is produced from pi.py (15):
    ranf function
 
Parallel for-loop #1: is produced from pi.py (15):
    array expression ((2 * '$0.6' (temporary variable)) - 1)
 
Parallel for-loop #2: is produced from pi.py (16):
    ranf function
 
Parallel for-loop #3: is produced from pi.py (16):
    array expression ((2 * '$0.15' (temporary variable)) - 1)
 
Parallel for-loop #4: is produced from pi.py (17):
    array expression (((x ** 2) + (y ** 2)) < 1)
 
Parallel for-loop #5: is produced from pi.py (17):
    numpy.sum (internal parallel version)
 
Found 1 parallel loops.
--------------------------------------------------------------------------------


Parallel loop listing for file pi.py
--------------------------------------------------------------------------------------|loop #ID
#                                                                                     | 
# Copyright (c) 2017 Intel Corporation                                                | 
# SPDX-License-Identifier: BSD-2-Clause                                               | 
#                                                                                     | 
                                                                                      | 
import numba                                                                          | 
import numpy as np                                                                    | 
import argparse                                                                       | 
import time                                                                           | 
                                                                                      | 
run_parallel = numba.config.NUMBA_NUM_THREADS > 1                                     | 
                                                                                      | 
@numba.njit(parallel=run_parallel)                                                    | 
def calc_pi(n):                                                                       | 
    x = 2*np.random.ranf(n)-1---------------------------------------------------------| #0, 1
    y = 2*np.random.ranf(n)-1---------------------------------------------------------| #2, 3
    return 4*np.sum(x**2+y**2<1)/n----------------------------------------------------| #4, 5
                                                                                      | 
def main():                                                                           | 
    parser = argparse.ArgumentParser(description='Calculate Pi.')                     | 
    parser.add_argument('--points', dest='points', type=int, default=20000000)        | 
    args = parser.parse_args()                                                        | 
    points = args.points                                                              | 
    np.random.seed(0)                                                                 | 
                                                                                      | 
    t1 = time.time()                                                                  | 
    pi = calc_pi(points)                                                              | 
    selftimed = time.time()-t1                                                        | 
    print("SELFTIMED ", selftimed)                                                    | 
    print("result: ", pi)                                                             | 
                                                                                      | 
if __name__ == '__main__':                                                            | 
    main()                                                                            | 
--------------------------------- Fusing loops ---------------------------------
Attempting fusion of parallel loops (combines loops with similar properties)...
  Trying to fuse loops #0 and #1:
    - fusion succeeded: parallel for-loop #1 is fused into for-loop #0.
  Trying to fuse loops #2 and #3:
    - fusion succeeded: parallel for-loop #3 is fused into for-loop #2.
  Trying to fuse loops #4 and #5:
    - fusion succeeded: parallel for-loop #5 is fused into for-loop #4.
  Trying to fuse loops #2 and #4:
    - fusion succeeded: parallel for-loop #4 is fused into for-loop #2.
  Trying to fuse loops #0 and #2:
    - fusion succeeded: parallel for-loop #2 is fused into for-loop #0.
 
Fused loop summary:
+--0 has the following loops fused into it:
   +--1 (fused)
   +--2 (fused)
      +--3 (fused)
      +--4 (fused)
         +--5 (fused)
Following the attempted fusion of parallel for-loops there are 1 parallel for-
loop(s) (originating from loops labelled: #0).
--------------------------------------------------------------------------------
----------------------------- Before Optimisation ------------------------------
Parallel region 0:
+--0 (parallel)
+--1 (parallel)
+--2 (parallel)
+--3 (parallel)
+--4 (parallel)
+--5 (parallel)


--------------------------------------------------------------------------------
------------------------------ After Optimisation ------------------------------
Parallel region 0:
+--0 (parallel, fused with loop(s): 1, 2, 3, 4, 5)


 
Parallel region 0 (loop #0) had 5 loop(s) fused.
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
 
---------------------------Loop invariant code motion---------------------------
Allocation hoisting:
No allocation hoisting found

Instruction hoisting:
loop #0 has the following hoisted:
  $0.2 = global(np: <module 'numpy' from '<path>/lib/python3.6/site-packages/numpy/__init__.py'>)
  $0.3 = getattr(value=$0.2, attr=random)
  $0.4 = getattr(value=$0.3, attr=ranf)
  $arg_out_var.46 = const(int, 2)
  $arg_out_var.48 = const(int, 1)
  $0.11 = global(np: <module 'numpy' from '<path>/lib/python3.6/site-packages/numpy/__init__.py'>)
  $0.12 = getattr(value=$0.11, attr=random)
  $0.13 = getattr(value=$0.12, attr=ranf)
  $arg_out_var.60 = const(int, 2)
  $arg_out_var.62 = const(int, 1)
  $arg_out_var.71 = const(int, 2)
  $arg_out_var.74 = const(int, 2)
  $arg_out_var.75 = const(int, 1)
--------------------------------------------------------------------------------
```
the diagnostics actually tracks internal parfors generation too, so the line:
```python
    x = 2*np.random.ranf(n)-1---------------------------------------------------------| #0, 1
```
has:
```
Parallel for-loop #0: is produced from pi.py (15):
    ranf function
 
Parallel for-loop #1: is produced from pi.py (15):
    array expression ((2 * '$0.6' (temporary variable)) - 1)
```
i.e. it saw the `ranf` function was transformed internally, and also that the array expression was a parfor.

-------------------------------------------------------------------------------------------------------------------------

Same applied to the logistic regression code, the key part (when `level=4`):
```
================================================================================
 Parallel Accelerator Optimizing:  Function logistic_regression, 
logistic_regression.py (9)  
================================================================================
---------------------------Looking for parallel loops---------------------------
Parallel for-loop #0: is produced from logistic_regression.py (12):
    array expression (-(Y))
 
Parallel for-loop #1: is produced from logistic_regression.py (12):
    array expression ((('$const12.5' (temporary variable) / ('$const12.6' 
(temporary variable) + (exp(('$12.10' (temporary variable) * '$12.15' (temporary
 variable)))))) - 1.0) * Y)
 
Parallel for-loop #2: is produced from logistic_regression.py (12):
    numpy.dot (internal parallel version)
 
Found 1 parallel loops.
--------------------------------------------------------------------------------


Parallel loop listing for file logistic_regression.py
-----------------------------------------------------------------------------------------|loop #ID
import numba                                                                             | 
import numpy as np                                                                       | 
import argparse                                                                          | 
import time                                                                              | 
                                                                                         | 
run_parallel = numba.config.NUMBA_NUM_THREADS > 1                                        | 
                                                                                         | 
                                                                                         | 
@numba.njit(parallel=run_parallel, fastmath=True)                                        | 
def logistic_regression(Y, X, w, iterations):                                            | 
    for i in range(iterations):                                                          | 
        w -= np.dot(((1.0 / (1.0 + np.exp(-Y * np.dot(X, w))) - 1.0) * Y), X)------------| #0, 1, 2
```
spots *a* `np.dot` is a parfor based on an internal parallel version, but doesn't declare which. I guess it'd be useful if I can work out a way to track that or trace it back to the call site (Python has no notion of source 'column' so it'll have to be based on the args to the function)?

**Comment by ehsantn** [comment]
I didn't know about level=4. This looks great! Tracking the actual call in the line is probably not essential at this point since shorter lines are more common. Printing input arguments to the function is useful though.

**Comment by DrTodd13** [comment]
Two thoughts.  Since the diagnostics call if just for one function, why print the entire file?  Can we just identify min and max line numbers for the functions IR and only print those lines?

Also, HPAT currently has some diagnostics and those are being reworked but we just refer to parfor regions by their lowest constituent parfor number.  These should match the parallel regions in these diagnostics so maybe take the same approach here?  Downside is that parallel region number will skip some numbers but that seems okay.

**Comment by stuartarchibald** [comment]
Marked as ready for review but as a feature it's very much experimental! Feedback/trying to break it most welcome, thanks.

File: numba/inline_closurecall.py [hunk 1]
```diff
@@ -510,17 +516,21 @@ def _find_iter_range(func_ir, range_iter_var):
     debug_print("func_var = ", func_var, " func_def = ", func_def)
     require(isinstance(func_def, ir.Global) and func_def.value == range)
     nargs = len(range_def.args)
+    swapping = [('"array comprehension"', 'closure of'), range_def.func.loc]
+    #[callname, repl_func.__name__, func_def, block.body[i].loc]
```
Comment (stuartarchibald) [inline numba/inline_closurecall.py]: rm this

File: numba/parfor.py [hunk 2]
```diff
@@ -15,8 +15,13 @@
 from __future__ import print_function, division, absolute_import
 import types as pytypes  # avoid confusion with numba.types
 import sys, math
+import os
+import linecache
```
Comment (stuartarchibald) [inline numba/parfor.py]: rm linecache, not used

File: numba/parfor.py [hunk 3]
```diff
@@ -576,17 +588,696 @@ def _analyze_parfor(parfor, equiv_set, typemap, array_analysis):
...
+        print_fusion_search = False
+        print_fusion_summary = False
+        print_loopnest_rewrite = False
+        print_pre_optimised = False
+        print_post_optimised = False
+        print_allocation_hoist = False
+        print_instruction_hoist = False
+        print_internal = False
+
+        if level == 1:
```
Comment (stuartarchibald) [inline numba/parfor.py]: could fold this

File: numba/compiler.py [hunk 4]
```diff
@@ -1015,14 +1036,15 @@ def type_inference_stage(typingctx, interp, args, return_type, locals={}):
 
 
 def native_lowering_stage(targetctx, library, interp, typemap, restype,
-                          calltypes, flags):
+                          calltypes, flags, parfor_diagnostics):
```
Comment (stuartarchibald) [inline numba/compiler.py]: Should this be the `metadata` container instead?

File: numba/tests/test_parfors.py [hunk 5]
```diff
@@ -333,7 +336,7 @@ def get_optimized_numba_ir(test_func, args, **kws):
         flags = compiler.Flags()
         parfor_pass = numba.parfor.ParforPass(
             tp.func_ir, tp.typemap, tp.calltypes, tp.return_type,
-            tp.typingctx, options, flags)
+            tp.typingctx, options, flags, diagnostics=diagnostics)
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: Also, add a few tests to run with diagnostics on.

File: numba/ir_utils.py [hunk 6]
```diff
@@ -1580,8 +1584,9 @@ def __init__(self, f_ir):
     rewrites.rewrite_registry.apply('before-inference',
                                     DummyPipeline(ir), ir)
     # call inline pass to handle cases like stencils and comprehensions
+    swapped = {} # TODO: get this from diagnostics store
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/ir_utils.py]: yes, it's not essential. If you happen to want some IR from a code object it's debatable as to whether tracking the swapped function in closure inlining is desirable, and if is it then the diagnostics store needs wiring in.

File: numba/inline_closurecall.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/inline_closurecall.py]: yes! done.

**Comment by seibert** [comment]
Seems to be [failing on 32-bit](https://dev.azure.com/numba/numba/_build/results?buildId=346&view=logs) due to range64 to 32 casting issues.  Example test fail:
```
2018-11-16T18:36:58.2368068Z ======================================================================
2018-11-16T18:36:58.2368125Z ERROR: test_alias_ctypes (numba.tests.test_remove_dead.TestRemoveDead)
2018-11-16T18:36:58.2368501Z ----------------------------------------------------------------------
2018-11-16T18:36:58.2368582Z Traceback (most recent call last):
2018-11-16T18:36:58.2368636Z   File "numba/tests/test_remove_dead.py", line 176, in test_alias_ctypes
2018-11-16T18:36:58.2368691Z     pfunc = self.compile_parallel(func, (numba.typeof(A1),))
2018-11-16T18:36:58.2368767Z   File "numba/tests/test_remove_dead.py", line 59, in compile_parallel
2018-11-16T18:36:58.2368835Z     return compile_isolated(func, arg_types, flags=fast_pflags).entry_point
2018-11-16T18:36:58.2368899Z   File "numba/compiler.py", line 158, in compile_isolated
2018-11-16T18:36:58.2368967Z     flags, locals)
2018-11-16T18:36:58.2369018Z   File "numba/compiler.py", line 926, in compile_extra
2018-11-16T18:36:58.2369069Z     return pipeline.compile_extra(func)
2018-11-16T18:36:58.2369140Z   File "numba/compiler.py", line 374, in compile_extra
2018-11-16T18:36:58.2369190Z     return self._compile_bytecode()
2018-11-16T18:36:58.2369242Z   File "numba/compiler.py", line 857, in _compile_bytecode
2018-11-16T18:36:58.2369310Z     return self._compile_core()
2018-11-16T18:36:58.2369362Z   File "numba/compiler.py", line 844, in _compile_core
2018-11-16T18:36:58.2369412Z     res = pm.run(self.status)
2018-11-16T18:36:58.2369481Z   File "numba/compiler_lock.py", line 32, in _acquire_compile_lock
2018-11-16T18:36:58.2369534Z     return func(*args, **kwargs)
2018-11-16T18:36:58.2369591Z   File "numba/compiler.py", line 255, in run
2018-11-16T18:36:58.2369640Z     raise patched_exception
2018-11-16T18:36:58.2369713Z LoweringError: Caused By:
2018-11-16T18:36:58.2369760Z Traceback (most recent call last):
2018-11-16T18:36:58.2369810Z   File "numba/compiler.py", line 246, in run
2018-11-16T18:36:58.2369875Z     stage()
2018-11-16T18:36:58.2369927Z   File "numba/compiler.py", line 717, in stage_nopython_backend
2018-11-16T18:36:58.2369979Z     self._backend(lowerfn, objectmode=False)
2018-11-16T18:36:58.2370046Z   File "numba/compiler.py", line 666, in _backend
2018-11-16T18:36:58.2370096Z     lowered = lowerfn()
2018-11-16T18:36:58.2370147Z   File "numba/compiler.py", line 653, in backend_nopython_mode
2018-11-16T18:36:58.2370218Z     self.metadata)
2018-11-16T18:36:58.2370270Z   File "numba/compiler.py", line 1048, in native_lowering_stage
2018-11-16T18:36:58.2370319Z     lower.lower()
2018-11-16T18:36:58.2370368Z   File "numba/lowering.py", line 177, in lower
2018-11-16T18:36:58.2370447Z     self.lower_normal_function(self.fndesc)
2018-11-16T18:36:58.2370501Z   File "numba/lowering.py", line 218, in lower_normal_function
2018-11-16T18:36:58.2370561Z     entry_block_tail = self.lower_function_body()
2018-11-16T18:36:58.2370630Z   File "numba/lowering.py", line 243, in lower_function_body
2018-11-16T18:36:58.2370681Z     self.lower_block(block)
2018-11-16T18:36:58.2370731Z   File "numba/lowering.py", line 258, in lower_block
2018-11-16T18:36:58.2370798Z     self.lower_inst(inst)
2018-11-16T18:36:58.2370855Z   File "/home/vsts/miniconda3/envs/travisci/lib/python2.7/contextlib.py", line 35, in __exit__
2018-11-16T18:36:58.2370912Z     self.gen.throw(type, value, traceback)
2018-11-16T18:36:58.2370980Z   File "numba/errors.py", line 609, in new_error_context
2018-11-16T18:36:58.2371033Z     six.reraise(type(newerr), newerr, tb)
2018-11-16T18:36:58.2371085Z   File "numba/errors.py", line 601, in new_error_context
2018-11-16T18:36:58.2371150Z     yield
2018-11-16T18:36:58.2371285Z   File "numba/lowering.py", line 258, in lower_block
2018-11-16T18:36:58.2371333Z     self.lower_inst(inst)
2018-11-16T18:36:58.2371432Z   File "numba/lowering.py", line 307, in lower_inst
2018-11-16T18:36:58.2371503Z     val = self.lower_assign(ty, inst)
2018-11-16T18:36:58.2371556Z   File "numba/lowering.py", line 453, in lower_assign
2018-11-16T18:36:58.2371606Z     return self.lower_expr(ty, value)
2018-11-16T18:36:58.2371677Z   File "numba/lowering.py", line 902, in lower_expr
2018-11-16T18:36:58.2371726Z     resty)
2018-11-16T18:36:58.2371775Z   File "numba/targets/base.py", line 682, in cast
2018-11-16T18:36:58.2371847Z     "Cannot cast %s to %s: %s" % (fromty, toty, val))
2018-11-16T18:36:58.2371910Z LoweringError: Cannot cast range_iter_int64 to range_iter_int32: %".159" = load {i64*, i64, i64, i64*}, {i64*, i64, i64, i64*}* %".116"
2018-11-16T18:36:58.2371950Z 
2018-11-16T18:36:58.2371998Z File "numba/tests/test_remove_dead.py", line 169:
2018-11-16T18:36:58.2372067Z         def func(ret):
2018-11-16T18:36:58.2372123Z             a = np.ones(4)
2018-11-16T18:36:58.2372170Z             ^
2018-11-16T18:36:58.2372216Z 
2018-11-16T18:36:58.2372281Z [1] During: lowering "$iter_var.662 = getiter(value=$range_c_var.661)" at numba/tests/test_remove_dead.py (169)
2018-11-16T18:36:58.2372320Z 
2018-11-16T18:36:58.2372375Z Cannot cast range_iter_int64 to range_iter_int32: %".159" = load {i64*, i64, i64, i64*}, {i64*, i64, i64, i64*}* %".116"
2018-11-16T18:36:58.2372433Z 
2018-11-16T18:36:58.2372751Z ----------------------------------------------------------------------
```

**Comment by seibert** [comment]
Ok, everything seems to be passing now!  Marking as ready to merge.  Documentation will come in a separate PR.

**Comment by seibert** [comment]
Merging now.  This is a really exciting feature that should hopefully demystify ParallelAccelerator a bit.  :)

**Pull request merged**